sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst

```diff
@@ -0,0 +1,162 @@
+C and C++ Standard libraries support
+===================================
+
+This extension enables a set of functions from the C and C++ standard
+libraries, and allows to use them in SYCL device code.
+
+Function declarations are taken from the standard headers (e.g. from
+<assert.h> or <cassert>), and the corresponding header has to be
+explicitly included in user code.
+
+Implementation requires a special device library to be linked with a
+SYCL program. The library should match the C or C++ standard library
+used to compile the program:
```

Comment:
No, see the line 77 in this document.
Basically, different libraries may use the same function name (e.g. __assert_fail), but with different arguments.

Reply:
I don't think it's the case for existing C/C++ standard library implementations.
Assuming that we want avoid potential issues with conflicting names, why do you propose "a special device library" instead of "a special header" as it's done for CUDA/OpenMP?

Reply:
> I don't think it's the case for existing C/C++ standard library implementations.

These little inconsistencies definitely exist.
For example, in musl C library, `__assert_fail` has an line argument of type `int`, and in Glibc has this argument as `unsigned int`.

Even if we consider only one library (e.g. Glibc), I'm not sure that ABI between headers and the library will always remain stable, so future/past versions may bring similar problems.

> Assuming that we want avoid potential issues with conflicting names, why do you propose "a special device library" instead of "a special header" as it's done for CUDA/OpenMP?

Can you describe this "special header" approach using the assert function as an example?

Reply:
Please, take a look here: https://github.com/intel/llvm/blob/sycl/clang/lib/Headers/__clang_cuda_runtime_wrapper.h#L335.
I suggest reading comments in https://github.com/intel/llvm/blob/sycl/clang/lib/Headers/__clang_cuda_* and take a look at https://github.com/intel/llvm/tree/sycl/clang/lib/Headers/cuda_wrappers.

Reply:
It is hard to tell how this approach is going to co-exist with the rest of the standard library, and scale beyond a handful of functions. The code you refer to doesn't seem to support anything other than glibc and libc++/libstdc++.

In general, I think the most difficult part here is (1) to establish an interface with a device compiler (DevicelibExtensions.rst), and (2) wire standard library functions through that interface, instead of using library specific routines.

(2) can probably be done in a header file, but I don't see a huge benefit in comparison with the library approach. If you (or anyone reading this) see a benefit and know to to implement this with portability in mind - feel free to send a patch.

Reply:
I think this at least should be discussed with the clang community.
It should be safe to develop solution aligned with the existing approaches in the community code w/o constantly checking with the community, but in this case we went in slightly different direction.